File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayDeque;
import java.util.Queue;
import java.util.StringTokenizer;

class Monkey{
	int x;
	int y;
	int cnt;
	int reK;
	
	public Monkey(int x, int y, int cnt, int reK) {
		this.x = x;
		this.y = y;
		this.cnt =cnt;
		this.reK = reK;
	}
}
class Main{
	
	static int K,W,H,res;
	static int[][] map;
	static int[] hdx = {-2,-1,1,2,2,1,-1,-2};
	static int[] hdy = {-1,-2,-2,-1,1,2,2,1};
	static int[] mdx = {0,1,0,-1};
	static int[] mdy = {1,0,-1,0};
	static boolean[][][] visited;
	static Queue<Monkey> q;

	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st;
		
		K = Integer.parseInt(br.readLine()); // 말의 이동 방법에서 K 횟수 가능
		
		st = new StringTokenizer(br.readLine());
		W = Integer.parseInt(st.nextToken()); // 가로
		H = Integer.parseInt(st.nextToken()); // 세로 
		
		map = new int[H][W];
		res = Integer.MAX_VALUE;
		visited = new boolean[K+1][H][W];
		
		for(int i=0;i<H;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<W;j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		} // 입력 완료 
		
		bfs();
		
		res = res != Integer.MAX_VALUE ? res : -1;
		// 출력
		bw.write(String.valueOf(res));
		bw.flush();
		br.close();
		bw.close();
		
	}
	
	static void bfs() {
		q = new ArrayDeque<Monkey>();
		
		// 출발 위치
		q.offer(new Monkey(0,0,0,K));
		visited[K][0][0] = true;
		
		while(!q.isEmpty()) {
			
			Monkey cur = q.poll();
			
			if(cur.x == H-1 && cur.y == W-1) {
				res = cur.cnt;
				return;
			}
			
			if(cur.reK > 0) {
				moveHorse(cur);
				moveMonkey(cur);
			}else {
				moveMonkey(cur);
			}
		}
		
	}
	
	static void moveHorse(Monkey cur) {
		
		for(int d=0;d<8;d++) {
			
			int nx = cur.x + hdx[d];
			int ny = cur.y + hdy[d];
			
			if(nx>-1 && nx<H && ny>-1 && ny<W) {
				if(visited[cur.reK-1][nx][ny] || map[nx][ny] == 1)  continue; // 방문했고 벽이면 안됨
				q.offer(new Monkey(nx,ny,cur.cnt+1, cur.reK-1));
				visited[cur.reK-1][nx][ny] = true;
			}
		}
	}
	
	static void moveMonkey(Monkey cur) {
		for(int d=0;d<4;d++) {
			
			int nx = cur.x + mdx[d];
			int ny = cur.y + mdy[d];
			
			if(nx>-1 && nx<H && ny>-1 && ny<W) {
				if(visited[cur.reK][nx][ny] || map[nx][ny] == 1)  continue; // 방문했고 벽이면 안됨
				
				q.offer(new Monkey(nx,ny,cur.cnt+1,cur.reK));
				visited[cur.reK][nx][ny] = true;
			}
		}
	}
}

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.cc
#include <iostream>
#include <queue>
using namespace std;

struct Info {
	int x;
	int y;
	int move_cnt;
	int jump_cnt;
};

int K, W, H; // 말처럼 움직일 수 있는 횟수, 가로 길이, 세로길이 
int map[201][201];
int dx[12] = { 0,0,1,-1,-2,-1,1,2,2,1,-1,-2 };
int dy[12] = { 1,-1,0,0,-1,-2,-2,-1,1,2,2,1 };
bool visited[31][201][201];

int BFS();

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	
	cin >> K >> W >> H;

	for (int i = 0;i < H;i++) {
		for (int j = 0;j < W;j++) {
			cin >> map[i][j];
		}
	}

	int res = BFS();

	cout << res << "\n";

	return 0;
}

int BFS() {
	queue<Info> q;
	q.push({ 0,0,0,0 });
	visited[0][0][0] = true;

	while (!q.empty()) {
		auto cur = q.front();
		q.pop();

		if (cur.x == H - 1 && cur.y == W - 1) return cur.move_cnt; // 도착점 도달 
		
		int nxt_x, nxt_y, cnt = cur.move_cnt + 1, jump = cur.jump_cnt;
		for (int d = 0;d < 12;d++) {
			if (d > 3) { // 말 점프 이동 경우
				if (cur.jump_cnt >= K) continue; // 말처럼 움직일 수 없는 경우
				jump = cur.jump_cnt + 1; // 점프 카운트 증가
			}
			
			nxt_x = cur.x + dx[d];
			nxt_y = cur.y + dy[d];

			if (nxt_x < 0 || nxt_x >= H || nxt_y < 0 || nxt_y >= W) continue; // 범위 초과 (먼저 처리해야함!)
			if (map[nxt_x][nxt_y] == 1) continue; // 다음 위치가 장애물인 경우 
			if (visited[jump][nxt_x][nxt_y]) continue; // 이미 방문한 경우 

			if (nxt_x == H - 1 && nxt_y == W - 1) return cnt; // 도착점 도달 


			q.push({nxt_x, nxt_y, cnt, jump});
			visited[jump][nxt_x][nxt_y] = true;
		}
	}

	return -1; // 시작점에서 도착점까지 이동 불가능 

}